gnomad/utils/constraint.py

```diff
@@ -13,6 +14,193 @@
+                freq_expr, freq_meta_expr, pop, singleton=True
+            )
+    # count the variants that have same combination of `grouping`
+    if use_table_group_by:
+        return ht.group_by(**grouping).partition_hint(partition_hint).aggregate(**agg)
+    else:
+        return ht.aggregate(
+            hl.struct(**{field: hl.agg.group_by(grouping, agg[field]) for field in agg})
+        )
+
+
+def downsampling_counts_expr(
+    freq_expr: hl.expr.ArrayExpression,
+    freq_meta_expr: hl.expr.ArrayExpression,
+    pop: str = "global",
+    variant_quality: str = "adj",
+    singleton: bool = False,
+    max_af: Optional[float] = None,
+) -> hl.expr.ArrayExpression:
+    """
+    Return an aggregation expression to compute an array of counts of all downsamplings found in `freq_expr` where specified criteria is met.
+
+    The meta data (`freq_meta_expr` annotated by annotate_freq) of each downsampling should have 'group',
+    'pop', and 'downsampling' keys. Included downsamplings are those where 'group' == `variant_quality` and 'pop' == `pop`.
+
+    :param freq_expr: ArrayExpression of Structs with with 'AC' and 'AF' annotations.
+    :param freq_meta_expr: ArrayExpression containing the set of groupings for each element of the `freq_expr` array (e.g., [{'group': 'adj'}, {'group': 'adj', 'pop': 'nfe'}, {'downsampling': '5000', 'group': 'adj', 'pop': 'global'}]).
+    :param pop: Population to use for filtering by the 'pop' key in `freq_meta_expr`. Defaults to 'global'.
+    :param variant_quality: Variant quality to use for filtering by the 'group' key in `freq_meta_expr`. Defaults to 'adj'.
+    :param singleton: Whether to filter to only singletons before counting (AC == 1). Defaults to False.
+    :param max_af: Maximum variant allele frequency to keep. By default no allele frequency cutoff is applied.
+    :return: Aggregation Expression for an array of the variant counts in downsamplings for specified population.
+    """
+    # get indices of dictionaries in meta dictionaries that only have the "downsampling" key with specified "group" and "pop" values
+    indices = hl.enumerate(freq_meta_expr).filter(
+        lambda f: (f[1].size() == 3)
+        & (f[1].get("group") == variant_quality)
+        & (f[1].get("pop") == pop)
+        & f[1].contains("downsampling")
+    )
+    # get an array of indices sorted by "downsampling" key
+    sorted_indices = hl.sorted(indices, key=lambda f: hl.int(f[1]["downsampling"])).map(
+        lambda x: x[0]
+    )
+
+    # variants met with specified criteria will be marked as 1
+    def _get_criteria(i: hl.expr.Int32Expression) -> hl.expr.Int32Expression:
+        if singleton:
+            return hl.int(freq_expr[i].AC == 1)
+        elif max_af:
+            return hl.int((freq_expr[i].AC > 0) & (freq_expr[i].AF <= max_af))
+        else:
+            return hl.int(freq_expr[i].AC > 0)
+
+    # mark variants met with specified criteria in each downsampling as 1
```

Comment:
Hmm, so the `hl.map` will apply  `_get_criteria` to each downsampling requested. so that will return a list of 1's and 0's for each variant, where if the variant exists in the downsampling (AC >0) and any specified criteria is met then it is given a 1 in the list, otherwise it is given a zero. 

Then the `hl.agg.array_sum` is a aggregation expression that when applied, will sum up all of these arrays of 1's and 0's to give an array that has one element per downsample in `pop` that is a count of variants that exist in the downsampling (AC >0) and, is requested, meet any specified criteria (`singleton` or `max_af`).

Reply:
A little bit more to clarify so I can think about a more accurate comment: I understand that this function only return an array sum aggregation. So after it's aggregated in the `count_variants_by_group()`, each element in the array is the sum of the count of variants that have the same `grouping` annotation right? And the variants mentioned in the previous sentence referred to those that exist in the downsampling (AC >0) and, is requested, meet any specified criteria (`singleton` or `max_af`).